File: infrastructure/src/main/java/com/study/platform/global/idempotency/IdempotencyStorageAdapter.java
package com.study.platform.global.idempotency;

import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Component;

import java.time.Duration;

@Component
@RequiredArgsConstructor
public class IdempotencyStorageAdapter implements IdempotencyStoragePort {

    private final StringRedisTemplate stringRedisTemplate;

    @Override
    public String get(String key) {
        return stringRedisTemplate.opsForValue().get(key);
    }

    @Override
    public boolean setIfAbsent(String key, String value, Duration ttl) {
        return Boolean.TRUE.equals(stringRedisTemplate.opsForValue().setIfAbsent(key, value, ttl));
    }

    @Override
    public void set(String key, String value, Duration ttl) {
        stringRedisTemplate.opsForValue().set(key, value, ttl);
    }

    @Override
    public void delete(String key) {
        stringRedisTemplate.delete(key);
    }
}

File: infrastructure/src/main/kotlin/com/study/platform/global/idempotency/IdempotencyStorageAdapter.kt
package com.study.platform.global.idempotency

import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.stereotype.Component
import java.time.Duration

@Component
class IdempotencyStorageAdapter(
    private val stringRedisTemplate: StringRedisTemplate
) : IdempotencyStoragePort {

    override fun get(key: String): String? =
        stringRedisTemplate.opsForValue().get(key)

    override fun setIfAbsent(key: String, value: String, ttl: Duration): Boolean =
        stringRedisTemplate.opsForValue().setIfAbsent(key, value, ttl) == true

    override fun set(key: String, value: String, ttl: Duration) {
        stringRedisTemplate.opsForValue().set(key, value, ttl)
    }

    override fun delete(key: String) {
        stringRedisTemplate.delete(key)
    }
}
